web/wp-content/plugins/tmbi-salesforce-dmp/class-tmbi-salesforce-dmp.php

```diff
@@ -0,0 +1,343 @@
+  }());" . PHP_EOL;
+
+		$contol_tag = sprintf( PHP_EOL . PHP_EOL . self::CONTROL_TAG_START . PHP_EOL . self::CONTROL_SCRIPT_TAG . PHP_EOL . self::CONTROL_TAG_END . PHP_EOL, $control_tag_script );
+
+		$interchage_tag_script = PHP_EOL . "window.Krux||((Krux=function(){Krux.q.push(arguments);}).q=[]);
+(function(){
+  function retrieve(n){
+    var k= 'kx'+'trustedmediabrandsinc_'+n, ls=(function(){
+      try {
+        return window.localStorage;
+      } catch(e) {
+        return null;
+      }
+    })();
+    if (ls) {
+        return ls[k] || '';
+    } else if (navigator.cookieEnabled) {
+        var m = document.cookie.match(k+'=([^;]*)');
+        return (m && unescape(m[1])) || '';
+    } else {
+        return '';
+    }
+  }
+  Krux.user = retrieve('user');
+  Krux.segments = retrieve('segs') ? retrieve('segs').split(',') : [];
+})();" . PHP_EOL;
+
+		$allowed_html   = [
+			'script' => [
+				'class'          => [],
+				'data-id'        => [],
+				'data-timing'    => [],
+				'data-version'   => [],
+				'type'           => [],
+				'data-namespace' => [],
+			],
+		];
+		$interchage_tag = sprintf( PHP_EOL . self::INTERCHANGE_SCRIPT_TAG, $interchage_tag_script );
+		$variant        = get_query_var( 'variant' );
+		if ( 'noads' !== $variant ) {
+			echo wp_kses( $contol_tag . $interchage_tag, $allowed_html );
+		}
+	}
+
+	/**
+	 * Function to build the krux data layer
+	 *
+	 * @return array
+	 */
+	public static function krux_build_datalayer() {
+
+		if ( ! is_single() ) {
+			return array();
+		}
+
+		$top_level_categories = '';
+		$sub_categories       = '';
+		$categories           = array();
+		$keywords             = '';
+
+		if ( class_exists( 'Ad_Stack' ) ) {
+			$categories = self::get_category_data();
+			$keywords   = self::get_post_tags();
+		}
+
+		$content_id = get_the_ID();
+		if ( is_array( $categories ) ) {
+			if ( array_key_exists( 'parent_categories', $categories ) && ( class_exists( 'Ad_Stack' ) ) ) {
+				$top_level_categories = self::format_data( $categories['parent_categories'] );
+			}
+			if ( array_key_exists( 'sub_categories', $categories ) && ( class_exists( 'Ad_Stack' ) ) ) {
+				$sub_categories = self::format_data( $categories['sub_categories'] );
+			}
+		}
+
+		self::$krux_data_layer['page'] = array(
+			'category' => $top_level_categories,
+			'topic'    => $sub_categories,
+			'keyword'  => $keywords,
+		);
+
+		self::$krux_data_layer['page']['contentid']     = $content_id;
+		self::$krux_data_layer['page']['author']        = self::get_post_authors();
+		self::$krux_data_layer['page']['gs_categories'] = json_decode( get_post_meta( $content_id, 'gs_channels', true ) );
+
+		$tax_array = self::get_post_taxonomy();
+		if ( $tax_array && ( class_exists( 'Ad_Stack' ) ) ) {
+			foreach ( $tax_array as $tax => $value ) {
+				self::$krux_data_layer['page'][ $tax ] = self::format_data( $value );
+			}
+		}
+		return self::$krux_data_layer;
```

Comment:
This function does too much and knows too much about its environment (e.g. the `gs_channels` meta field, getting info from taxonomies, tags, etc). Consider breaking it down in smaller functions, and collect the results from each of them via filters.